### PR TITLE
Epic refactoring of DockerHelper and friends

### PR DESCRIPTION
**Pull request by jerith**

### CODE REVIEW

File: seaworthy/containers/base.py [hunk 1]
```diff
@@ -50,28 +62,82 @@ def __init__(self, name, image, wait_patterns=None, wait_timeout=None,
...
+        self._teardown()
+
+    def _teardown(self):
+        """
+        Stop and remove the container if it exists.
+        """
+        if self._container is not None:
+            self.stop_and_remove()
+
+    def as_fixture(self, name=None):
```
Comment (JayH5) [inline seaworthy/containers/base.py]: Docs please. I don't really understand what's happening here 🙃.

File: seaworthy/containers/base.py [hunk 2]
```diff
@@ -50,28 +62,82 @@ def __init__(self, name, image, wait_patterns=None, wait_timeout=None,
...
+        if docker_helper is None:  # We don't want to "unset" in this method.
+            return
+        if docker_helper is self._docker_helper:  # We already have this one.
+            return
+        if self._docker_helper is None:
+            self._docker_helper = docker_helper
+        else:
+            raise RuntimeError('Cannot replace existing docker_helper.')
+
+    def status(self):
```
Comment (JayH5) [inline seaworthy/containers/base.py]: Is this just for tests? Should we maybe mention that this does a reload before getting the status and so should return a reasonably up-to-date status?

File: seaworthy/containers/base.py [hunk 3]
```diff
@@ -50,28 +62,82 @@ def __init__(self, name, image, wait_patterns=None, wait_timeout=None,
...
+        """
+        Get the container's status. If the container does not exist (before
+        creation and after removal), the status is ``None``.
+        """
+        if self._container is None:
+            return None
+        return self.docker_helper.containers.status(self.inner())
+
+    def create_and_start(self, docker_helper=None, pull=True, kwargs=None):
```
Comment (JayH5) [inline seaworthy/containers/base.py]: I'm thinking we should rename the `run`.. as that's basically what it is. But that's a separate PR, probably.

File: seaworthy/containers/base.py [hunk 4]
```diff
@@ -50,28 +69,104 @@ def __init__(self, name, image, wait_patterns=None, wait_timeout=None,
...
+    def status(self):
+        """
+        Get the container's current status from docker.
+
+        If the container does not exist (before creation and after removal),
+        the status is ``None``.
+        """
+        if self._container is None:
+            return None
+        return self.docker_helper.containers.status(self.inner())
```
Comment (JayH5) [inline seaworthy/containers/base.py]: This logic I think would be better in ContainerBase, like:
```python
inner = self.inner()
inner.reload()
return inner.status
```
or whatever.. but I guess that can be in another PR.

Comment (jerith) [reply seaworthy/containers/base.py]: That's a small enough change that it's worth doing here. :-)
